clang/lib/CIR/CodeGen/CIRGenModule.cpp

```diff
@@ -24,9 +27,140 @@ CIRGenModule::CIRGenModule(mlir::MLIRContext &context,
                            clang::ASTContext &astctx,
                            const clang::CodeGenOptions &cgo,
                            DiagnosticsEngine &diags)
-    : astCtx(astctx), langOpts(astctx.getLangOpts()),
-      theModule{mlir::ModuleOp::create(mlir::UnknownLoc())},
-      target(astCtx.getTargetInfo()) {}
+    : builder(&context), astCtx(astctx), langOpts(astctx.getLangOpts()),
+      theModule{mlir::ModuleOp::create(mlir::UnknownLoc::get(&context))},
+      diags(diags), target(astCtx.getTargetInfo()) {}
+
+mlir::Location CIRGenModule::getLoc(SourceLocation cLoc) {
+  assert(cLoc.isValid() && "expected valid source location");
+  const SourceManager &sm = astCtx.getSourceManager();
+  PresumedLoc pLoc = sm.getPresumedLoc(cLoc);
+  StringRef filename = pLoc.getFilename();
+  return mlir::FileLineColLoc::get(builder.getStringAttr(filename),
+                                   pLoc.getLine(), pLoc.getColumn());
+}
```

Comment:
I am guessing that the MLIR location type doesn't have any notion of multiple nested locations due to macro expansions.  So we can't represent the full complexity of macros here.  The outermost location seems like the right one to choose.  (I am not familiar with either the Clang location type or the MLIR location type, so I welcome being corrected here.)

Reply:
I'll leave this open for Aaron to comment on.  IIRC, the source location stored in the AST is going to be the location of the macro, which can make diagnostics awkward/confusing if what happens is inside the expansion, but if MLIR can't comprehend "in expansion of..." then this is perhaps as good as we can do.

Reply:
I think this may be okay -- a `PresumedLoc` is sensitive to `#line` and GNU line markers, etc, and it's always the expansion point of the wrapped source location. However, MLIR may eventually need to grow to understand the difference between spelling, expansion, and presumed locations.

Reply:
If you ever want to generate diagnostics from CIR, you probably want to just use the clang SourceLocation directly, not try to translate it to an MLIR "Location".  It's extremely complicated to describe a location when macros are involved, and the abstraction clang uses really only makes sense for C preprocessor.

Reply:
Diagnostics that happen during CIR code generation will use the Clang `SourceLocation` because the Clang AST is still available.  But later passes won't have access to a `SourceLocation`.  Every MLIR node already has an `mlir::Location`.  Storing a `SourceLocation` instead of, or in addition to, an `mlir::Location` is not practical.

I don't expect many diagnostics to be issued from CIR when compilation.  Virtually all diagnostics of interest should have been issued during parsing and semantic checking.  The answer may be different when CIR is used for static analysis.  I'll let the static analysis experts sort that out.

Reply:
> I don't buy the reasoning of "mlir::Location exists, therefore it's the right representation".

`mlir::Location` is not the right representation because it exists.  It's because it is used by all the other MLIR dialects that have nothing to do with C++ and cannot have a dependency on Clang.  It is wanting ClangIR to play nicely with the rest of the MLIR ecosystem, when many dialects are mixed together in a single tree later in the compilation chain, that makes using `SourceLocation` problematic.

> Not saying this is something we need to solve right now

Definitely.  Using `SourceLocation` instead of `mlir::Location` is not something that can be done in this PR.  We would need evidence that the lack of `SourceLocation` and the richness of what it can represent is a serious problem before attempting such a radical change.

Reply:
>  It is wanting ClangIR to play nicely with the rest of the MLIR ecosystem, when many dialects are mixed together in a single tree later in the compilation chain, that makes using SourceLocation problematic.

However, that's also an internal implementation detail that doesn't need to be exposed to users of the CIR library. Clang works in `SourceLocation`; if CIR needs to work in something else, that should be abstracted away from the user. Still not saying this needs to be solved as part of this patch, but `CIRGenModule::getLoc()` are public interfaces currently; can those be made private so we're not leaking an implementation detail of CIR to users?

> But later passes won't have access to a SourceLocation.

Two things:

1) Diagnostics that depend on optimization levels have massive usability issues, so diagnostics from later passes strikes me as a potential design concern to start with; hopefully we don't end up in that situation except in very rare circumstances.
2) The backend does occasionally need to issue user-facing diagnostics (things like `__attribute__((error))` come to mind) and this has been a huge source of pain because LLVM does a pretty bad job of tracking source location with sufficient fidelity for Clang to have good QoI. If MLIR has the same issues, that should be something the MLIR folks work to correct before we transition to CIR as a default.

Reply:
> However, that's also an internal implementation detail that doesn't need to be exposed to users of the CIR library. 

The primary use of ClangIR is intended to be of the MLIR data structures after CIR code gen has happened.  The public interface for ClangIR is entirely based on MLIR, including `mlir::Location`.  Some users will want to use the CIR library to convert from the Clang AST to the ClangIR MLIR dialect, but for many users of ClangIR the fact that the MLIR data structures were generated from the Clang front end and the Clang AST is an implementation detail that they don't need to know about.  `SourceLocation` is more of an implementation detail than `mlir::Location` is.

Changing `CIRGenModule::getLoc()` to be private functions would be problematic because those functions will be used by many different classes within the ClangIR code gen code.  Those functions really should have module accessibility (similar to package or assembly accessibility in Java or C#), but that's not practical until Clang/LLVM fully moves to C++20 modules.

I am aware and agree that (1) back end diagnostics should be avoided where possible and should be rare, but that (2) when they do happen they can be hard to get right.  We struggle with that same problem with NVC++.  We should solve the problem of getting good diagnostics during the MLIR passes when we actually run into the problem.  It is only then that we will know how serious the problem is, how best to solve it, and how much effort to put into it.  Solving that problem now is premature and will miss the mark in some way.

Reply:
>  SourceLocation is more of an implementation detail than mlir::Location is.

That depends on perspective, IMO. From our perspective, CIR is the way in which we lower Clang's AST to an IR that eventually ends up in LLVM. The public entrypoints into CIR from Clang's perspective should be using Clang's data structures. If those need to be converted internally into MLIR-specific data structures, that's fine, but that should not leak out into the public interfaces that Clang can interact with.

However, I see now that I missed something important here -- this code is under `clang/lib/CIR` and not `clang/include/clang/CIR`, so it *isn't* in the public interfaces that Clang interacts with, it is (sufficiently) hidden as an implementation detail. Sorry for missing that!

> We should solve the problem of getting good diagnostics during the MLIR passes when we actually run into the problem. It is only then that we will know how serious the problem is, how best to solve it, and how much effort to put into it. Solving that problem now is premature and will miss the mark in some way.

The point to these initial PRs is to ensure the community is comfortable with the design and doesn't see any glaring design concerns, so saying "we'll solve it when we get to it" on a situation we know we will hit sort of defeats the purpose. We already know the problem exists because it's one we fight with today and we already know a better way to solve it is to use source location information with at least as much fidelity as Clang is able to handle. Given that one of the benefits to CIR is for better codegen analysis that can lead to improved diagnostics, having this as an early design concern is pretty reasonable IMO. That said, we're not insisting on a change as part of this PR -- just that the CIR folks understand that this is a design concern with the facilities and at some point it may go from "concern" to "required to improve before we can proceed."